beacon-chain/rpc/prysm/v1alpha1/validator/unblinder.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/runtime/version"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+)
+
+type unblinder struct {
+	b       interfaces.SignedBeaconBlock
+	builder builder.BlockBuilder
+}
+
+func newUnblinder(b interfaces.SignedBeaconBlock, builder builder.BlockBuilder) (*unblinder, error) {
+	if err := consensusblocks.BeaconBlockIsNil(b); err != nil {
+		return nil, err
+	}
+	return &unblinder{
+		b:       b,
+		builder: builder,
+	}, nil
+}
+
+func (u *unblinder) unblindBuilderBlock(ctx context.Context) (interfaces.SignedBeaconBlock, error) {
+	if !u.b.IsBlinded() || u.b.Version() < version.Bellatrix {
+		return u.b, nil
+	}
+	if u.b.IsBlinded() && !u.builder.Configured() {
+		return nil, errors.New("builder not configured")
+	}
+
+	agg, err := u.b.Block().Body().SyncAggregate()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get sync aggregate")
+	}
+	h, err := u.b.Block().Body().Execution()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get execution")
+	}
+	parentRoot := u.b.Block().ParentRoot()
+	stateRoot := u.b.Block().StateRoot()
+	randaoReveal := u.b.Block().Body().RandaoReveal()
+	graffiti := u.b.Block().Body().Graffiti()
+	sig := u.b.Signature()
+	blsToExecChanges, err := u.b.Block().Body().BLSToExecutionChanges()
+	if err != nil && !errors.Is(err, consensusblocks.ErrUnsupportedGetter) {
+		return nil, errors.Wrap(err, "could not get bls to execution changes")
+	}
+
+	psb, err := u.blindedProtoBlock()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get blinded proto block")
+	}
+	sb, err := consensusblocks.NewSignedBeaconBlock(psb)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create signed block")
+	}
+	sb.SetSlot(u.b.Block().Slot())
+	sb.SetProposerIndex(u.b.Block().ProposerIndex())
+	sb.SetParentRoot(parentRoot[:])
+	sb.SetStateRoot(stateRoot[:])
+	sb.SetRandaoReveal(randaoReveal[:])
+	sb.SetEth1Data(u.b.Block().Body().Eth1Data())
+	sb.SetGraffiti(graffiti[:])
+	sb.SetProposerSlashings(u.b.Block().Body().ProposerSlashings())
+	sb.SetAttesterSlashings(u.b.Block().Body().AttesterSlashings())
+	sb.SetAttestations(u.b.Block().Body().Attestations())
+	sb.SetDeposits(u.b.Block().Body().Deposits())
+	sb.SetVoluntaryExits(u.b.Block().Body().VoluntaryExits())
+	if err = sb.SetSyncAggregate(agg); err != nil {
+		return nil, errors.Wrap(err, "could not set sync aggregate")
+	}
+	sb.SetSignature(sig[:])
+	if err = sb.SetBLSToExecutionChanges(blsToExecChanges); err != nil && !errors.Is(err, consensusblocks.ErrUnsupportedGetter) {
+		return nil, errors.Wrap(err, "could not set bls to execution changes")
+	}
+	if err = sb.SetExecution(h); err != nil {
+		return nil, errors.Wrap(err, "could not set execution")
+	}
+
+	payload, err := u.builder.SubmitBlindedBlock(ctx, sb)
```

Comment:
It might be a good time to add a defensive check for SubmitBlindedBlock and do a version check on the response for the payload. right now the version isn't getting checked on the response and is not part of the executiondata interface which could result in a bad payload.

Reply:
yeah we need a new PR for checking inside I'm not sure if ExecutionData needs it to get the version or not, but at least it should be checked inside

Reply:
my PR merged in with the check here